core/primitives/src/runtime/parameter_table.rs

```diff
@@ -16,36 +16,90 @@ pub(crate) enum ParameterValue {
+            _ => Err(ValueConversionError::WrongType(std::any::type_name::<u64>(), value.clone())),
         }
     }
+}
 
-    fn as_rational(&self) -> Option<Rational32> {
-        match self {
-            ParameterValue::Rational { numerator, denominator } => {
-                Some(Rational32::new(*numerator, *denominator))
-            }
-            _ => None,
+impl TryFrom<&ParameterValue> for u128 {
+    type Error = ValueConversionError;
+
+    fn try_from(value: &ParameterValue) -> Result<Self, Self::Error> {
+        match value {
+            ParameterValue::U64(v) => Ok(u128::from(*v)),
+            ParameterValue::String(s) => s.parse().map_err(|err| {
+                ValueConversionError::ParseInt(err, std::any::type_name::<u128>(), value.clone())
+            }),
+            _ => Err(ValueConversionError::WrongType(std::any::type_name::<u128>(), value.clone())),
         }
     }
+}
 
-    fn as_u128(&self) -> Option<u128> {
-        match self {
-            ParameterValue::U64(v) => Some(u128::from(*v)),
-            // TODO(akashin): Refactor this to use `TryFrom` and properly propagate an error.
-            ParameterValue::String(s) => s.parse().ok(),
-            _ => None,
+// TODO(akashin): Enable this.
+// impl<T> TryFrom<&ParameterValue> for T
+// where
+//     T: TryFrom<u64>,
+//     T::Error: Into<std::num::TryFromIntError>,
+// {
+//     type Error = ValueConversionError;
+//
+//     fn try_from(value: &ParameterValue) -> Result<Self, Self::Error> {
+//         match value {
+//             ParameterValue::U64(v) => T::try_from(*v).map_err(|err| {
+//                 ValueConversionError::TryFromInt(
+//                     err.into(),
+//                     std::any::type_name::<T>(),
+//                     value.clone(),
+//                 )
+//             }),
+//             _ => Err(ValueConversionError::WrongType(std::any::type_name::<T>(), value.clone())),
+//         }
+//     }
+// }
+
+impl TryFrom<&ParameterValue> for Rational32 {
+    type Error = ValueConversionError;
+
+    fn try_from(value: &ParameterValue) -> Result<Self, Self::Error> {
+        match value {
+            ParameterValue::Rational { numerator, denominator } => {
+                Ok(Rational32::new(*numerator, *denominator))
+            }
+            _ => Err(ValueConversionError::WrongType(
+                std::any::type_name::<Rational32>(),
+                value.clone(),
+            )),
         }
     }
+}
+
+impl TryFrom<&ParameterValue> for String {
+    type Error = ValueConversionError;
 
-    fn as_str(&self) -> Option<&str> {
-        match self {
-            ParameterValue::String(v) => Some(v),
-            _ => None,
+    fn try_from(value: &ParameterValue) -> Result<Self, Self::Error> {
+        match value {
+            ParameterValue::String(v) => Ok(v.clone()),
+            _ => {
+                Err(ValueConversionError::WrongType(std::any::type_name::<String>(), value.clone()))
```

Comment:
I see your point about shorter/simpler error message being more readable, but on the other hand I feel this will introduce the inconsistency in the code (using different ways to derive this name) and also in the error message itself (mixing different ways to write type names).

Overall, both approaches are fine with me, so happy to go with whatever you think is more consistent with existing codebase/Rust codestyle.

Reply:
Given that this is not a deal-breaker and both approaches have merit, I'll leave things as is (consistent with original code that I was modifying). Happy to revisit this in the future in a separate PR.